labs/projecteuler/projecteuler.md

```diff
@@ -0,0 +1,132 @@
+# Properties in Cryptol, or How I Learned to Stop Worrying and Love SAT Solvers
+
+You have learned in the other labs that you can use Cryptol to find answers to problems you coded using the `:sat` command.  As a refresher, let's look at a simple example.  Suppose we want to find the integer that is one less than 1001 (as we said, it's very simple).  We do this by creating a function and a `property`:
+
+```
+inc : Integer -> Integer
+inc x = x + 1
+
+inc1001 : Integer -> Bit
+property inc1001 x = inc x == 1001
+```
+
+Now that we have our function and a property about our function, we can load it into Cryptol and find an answer!
+
+```shell
+:sat inc1001
+```
+You should get a response like this:
+```shell
+Main> :sat inc1001 
+inc1001 1000 = True
+(Total Elapsed Time: 0.039s, using Z3)
+```
+Let's do a more complicated example: use Cryptol to factor 3,000,013.
+
+```
+mult : Integer -> Integer -> Integer
+mult x y = x * y
+
+mult3000013 : Integer -> Integer -> Bit
+mult3000013 x y =
+    x * y == 3000013 /\
+    x > 1            /\
+    y > 1
+```
+Note that if we don't include the `x != 1` clause we get a trivial factorization.  Now we can use SAT to factor our number:
+```shell
+Main> :sat mult3000013 
+mult3000013 773 3881 = True
+(Total Elapsed Time: 0.684s, using Z3)
+```
+Clearly we don't need SAT solvers to figure out how to subtract one.  And writing simple factorization code in Python factored 3,000,013 in 0.419s.  But there is something really interesting here: to solve the problem, instead of figuring out how to write code which would solve the problem for us, we just had to write the problem in Cryptol code and let a SAT solver solve it for us!  Usually it's much easier to write a problem than write the solution.  We can't always expect the SAT solver to find the answer quickly (this is known as as the [P vs. NP problem](https://en.wikipedia.org/wiki/P_versus_NP_problem)), but there are many times where it will come back with answer quickly.  We like to say that it doesn't cost much to ask a SAT solver.
+
+## Project Euler Problems
+
+[Project Euler](https://projecteuler.net/) is a collection of mathematical problems that can be solved using relatively simple computer code (i.e. to do them, you shouldn't need to create optimized code to run on a supercomputer).  Some of the questions lend themselves to querying a SAT solver.  We've collected them here for you to try your hand at.
```

Comment:
Suggest "... (i.e., the solution shouldn't require optimized code running on an supercomputer)..."